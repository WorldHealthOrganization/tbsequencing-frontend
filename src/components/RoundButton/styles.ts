import { css, SerializedStyles } from '@emotion/react';
import appColors from '../../styles/colors';

const XLARGE = 72;
const LARGE = 64;
const NORMAL = 56;
const SMALL = 40;
const XSMALL = 32;

const createBtnStyle = (size: number): SerializedStyles => css({
  width: size,
  height: size,
  borderRadius: size / 2,
  backgroundColor: appColors.secondary.whoBlue,
  color: appColors.neutral.white,
  '&:hover': {
    boxShadow: '0px 1px 1px rgba(0, 0, 0, 0.2), 0px 2px 2px rgba(0, 0, 0, 0.14), 0px 1px 5px rgba(0, 0, 0, 0.12)',
    backgroundColor: appColors.secondary.whoBlue,
  },
});

export const xlarge = createBtnStyle(XLARGE);
export const large = createBtnStyle(LARGE);
export const normal = createBtnStyle(NORMAL);
export const small = createBtnStyle(SMALL);
export const xsmall = createBtnStyle(XSMALL);
