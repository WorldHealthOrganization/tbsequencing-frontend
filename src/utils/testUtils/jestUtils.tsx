import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
