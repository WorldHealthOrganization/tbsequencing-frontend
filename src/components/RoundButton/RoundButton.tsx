/** @jsxImportSource @emotion/react */
import React from 'react';
import { SerializedStyles } from '@emotion/react';
import { match } from 'ts-pattern';
import { IconButton } from '@mui/material';
import { Add, Remove } from '@mui/icons-material';
import * as styles from './styles';

export type IButtonSize = 'xlarge' | 'large' | 'normal' | 'small' | 'xsmall';
export type IButtonVariants = 'add' | 'remove';

interface Props {
  style?: SerializedStyles;
  size: IButtonSize;
  variant: IButtonVariants;
  disabled?: boolean;
  onClick?: () => void;
}

export const RoundButton = ({
  style, size, variant, disabled, onClick,
}: Props) => {
  const isDisabled = Boolean(disabled);

  const icon = match(variant)
    .with('add', () => <Add />)
    .with('remove', () => <Remove />)
    .exhaustive();

  const sizeStyle = match(size)
    .with('xlarge', () => styles.xlarge)
    .with('large', () => styles.large)
    .with('normal', () => styles.normal)
    .with('small', () => styles.small)
    .with('xsmall', () => styles.xsmall)
    .exhaustive();

  return (
    <IconButton
      onClick={onClick}
      disabled={isDisabled}
      css={[sizeStyle, style]}
    >
      {icon}
    </IconButton>
  );
};

export default RoundButton;
