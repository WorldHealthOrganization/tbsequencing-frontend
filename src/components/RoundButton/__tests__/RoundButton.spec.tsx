import React from 'react';
import { getTreeJSONWithRedux, render } from '../../../utils/testUtils/jestUtils';
import RoundButton from '../RoundButton';

describe('RoundButton component test', () => {
  const mockOnClick = jest.fn();

  test('renders RoundButton component without error', () => {
    render(
      <RoundButton onClick={mockOnClick} size="small" variant="add" />,
    );
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <RoundButton onClick={mockOnClick} size="small" variant="add" />,
    );

    expect(tree).toMatchSnapshot();
  });
});
