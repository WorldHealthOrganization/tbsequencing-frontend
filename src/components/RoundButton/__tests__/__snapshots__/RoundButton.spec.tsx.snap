// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoundButton component test Should render correctly 1`] = `
<button
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-pgl9qq-MuiButtonBase-root-MuiIconButton-root-RoundButton"
  disabled={false}
  onBlur={[Function]}
  onClick={[MockFunction]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
    data-testid="AddIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
    />
  </svg>
</button>
`;
